### ISSUE TITLE
Regression: Linearly-increasing CPU usage as time goes on starting in 0.30

### ISSUE THREAD

**Issue by nickwilliams-eventbrite** [DataDog/dd-trace-py#1114]
Starting in 0.30, in applications with very high load and very little idle time, the `ddtrace` writer thread uses higher CPU that increases linearly as time elapses from application startup. It took us the better part of two days to figure out this was the culprit, but we have no idea why. The problem manifested when other metrics (not tracing) showed us that 1) the CPU the application was using kept increasing linearly as the application continued to run, and 2) blocking I/O processes in the application kept taking longer and longer as the application continued to run. When we downgraded to 0.29.0, the problem went away completely.

The first sign of trouble was blocking I/O processes in the application taking longer and longer, and then we noticed the CPU increasing perpetually. I/O processes that normally take ~1ms would eventually take 30ms after about 3 hours of application runtime and then reset to 1ms and start growing again when the application was restarted. We are getting these I/O times from independent timers (_not_ using `ddtrace`) that were publishing to DogStatsd. It didn't make sense to us that different, unrelated I/O processes (Redis, MySQL, etc.) were taking longer, so we suspected that the underlying I/O actually was _not_ degrading, but that what was _actually_ happening was that a background thread of some type was binding up the CPU while the main thread was blocked waiting on I/O, and that the background thread took so long that the main thread observed longer I/O times when really it was just waiting for the background thread to yield. Our application had only two background threads: The `ddtrace` writer and a Sentry writer, and the Sentry client had not been recently upgraded, so we deduced the problem must be `ddtrace`. Downgrading to 0.29.0 and observing that the problem with away confirmed this deduction.

So something in the `ddtrace` background thread that publishes traces to the agent is using abnormal amounts of CPU, and the amount of CPU it uses increases continuously the longer the application runs (I suspect it's holding on to something and not flushing it properly). This is resulting in overall degraded application performance due to both the increased CPU usage and the Python Global Interpreter Lock. Note that our application's tracing did not look unusual ... we seemed to be getting the same, normal traces that we were getting before.

**Comment by majorgreys** [comment]
@nickwilliams-eventbrite Thank you for raising as an issue. Before we dig further, I wanted to make sure you are on the current release, 0.30.2, which includes a few bug fixes we identified in 0.30.0.

**Comment by nickwilliams-eventbrite** [comment]
With much caution and anxiety, we tested 0.31.0 in Production today, and it no longer exhibits the symptoms observed in 0.30.x. I would really love for us all to understand better what caused this issue and what fixed it, so that we keep it from happening again in the future, but for now this issue is resolved.

**Comment by majorgreys** [comment]
I appreciate your team upgrading to 0.31.0 after the problems with 0.30.x and I'm glad it's working out. I'll close this issue for now, but please re-open it if anything crops up again with cpu usage.

**Comment by brettlangdon** [comment]
@nickwilliams-eventbrite if you were using `0.30.0` we accidentally released with some new metrics collection enabled by default. As well, this version of our metrics collection included calculating `sizeof` of every trace we created, this in particular we found to have a negative performance impact.

In later bug fix versions of `0.30.x` we disabled these metrics by default, and in the latest version of `0.31.0` (still disabled by default) we removed this heavy metric all together.
